traits/trait_types.py

```diff
@@ -2774,6 +2774,178 @@ def items_event(self):
 AdaptMap = {"no": 0, "yes": 1, "default": 2}
 
 
+class Map(TraitType):
+    """ Checks that the value assigned to a trait attribute is a key of a
+        specified dictionary, and also assigns the dictionary value
+        corresponding to that key to a *shadow* attribute.
+
+        A trait attribute of type Map is called a *mapped* trait
+        attribute. In practice, this means that the resulting object actually
+        contains two attributes: one whose value is a key of the Map
+        dictionary, and the other whose value is the corresponding value of the
+        Map dictionary. The name of the shadow attribute is simply the base
+        attribute name with an underscore ('_') appended. Mapped trait
+        attributes can be used to allow a variety of user-friendly input values
+        to be mapped to a set of internal, program-friendly values.
+
+        Example
+        -------
+
+        The following example defines a ``Person`` class::
+
+            >>>class Person(HasTraits):
+            ...    married = Map({'yes': 1, 'no': 0 })
+            >>>
+            >>>bob = Person()
+            >>>print bob.married
+            yes
+            >>>print bob.married_
+            1
+
+        In this example, the default value of the ``married`` attribute of the
+        Person class is 'yes'. Because this attribute is defined using
+        PrefixList, instances of Person have another attribute,
+        ``married_``, whose default value is 1, the dictionary value
+        corresponding to the key 'yes'.
+
+        Parameters
+        ----------
+        map : dict
+            A dictionary whose keys are valid values for the trait attribute,
+            and whose corresponding values are the values for the shadow
+            trait attribute.
+
+        Attributes
+        ----------
+        map : dict
+            A dictionary whose keys are valid values for the trait attribute,
+            and whose corresponding values are the values for the shadow
+            trait attribute.
+    """
+
+    is_mapped = True
+
+    def __init__(self, map, **metadata):
+
+        self.map = map
+        self.fast_validate = (ValidateTrait.map, map)
+
+        default_value = metadata.pop("default_value", None)
```

Comment:
Use `Undefined` as the default default value?

Reply:
I'm not sure of the convention followed in `traits` . When do we use `Undefined` as opposed to None?

Reply:
If a trait value can be None, we need something that is not `None` to signal when we do not know the value.  This is different from knowing the value of trait being None.

In this case, `None` is hashable and can be used as a key in `map`, i.e. it can be a valid value.
Also, since mapping has no ordering, we cannot reasonably assume any default default value.

The `default_value` in `TraitType` is `Undefined`:
https://github.com/enthought/traits/blob/1d7645755d2811fe7c8c155c6ff85adbe89107c3/traits/trait_type.py#L179-L180

(Orthogonal: I am not a fan of many of defaults in `trait_types`. They strike me as something motivated by GUI, which should be dealt with separately in the GUI context.)